File: data-structures/queues/python/queue.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


MAX_SIZE = 4 # Max size of the queue


def is_empty(front, rear):
    """Check if the queue is empty"""
    
    if front > rear:
        return True
    else:
        return False


def is_full(rear):
    """Check if the queue is full"""
    
    if rear + 1 == MAX_SIZE:
        return True
    else:
        return False


def enqueue(queue, rear, data):
    """Enqueue an item"""
    
    if is_full(rear) == True:
        print(f"\nQueue is full - {data} not added")
    else:          
        rear = rear + 1
        queue[rear] = data
    return rear


def dequeue(queue, front, rear):
    """Dequeue an item"""
    
    if is_empty(front, rear) == True:
        print("\nQueue is empty - nothing to dequeue")
        dequeued_item = None
    else:
        dequeued_item = queue[front]
        front = front + 1
    return dequeued_item, front


def output_queue(queue, front, rear):
    """Output the state of the queue"""

    # Testing
    print(f"\nQueue: {queue}")
    print(f"Front: {front}")
    print(f"Rear: {rear}")

        
def main():
    """Create a queue and enqueue and dequeue items"""
    
    print("### Queue ###")

    # Initialise the queue and pointers
    queue = []
    front = 0
    rear = -1

    # Simulate a fixed size array
    for i in range(MAX_SIZE):
        queue.append(None)
    output_queue(queue, front, rear)

    # Insert test data into the queue
    rear = enqueue(queue, rear, "Julie")
    output_queue(queue, front, rear)
    
    rear = enqueue(queue, rear, "Rey")
    output_queue(queue, front, rear)
    
    rear = enqueue(queue, rear, "Habib")
    output_queue(queue, front, rear)
    
    rear = enqueue(queue, rear, "Sabrina")
    output_queue(queue, front, rear)

    # Trying to enqueue - should say queue is full
    rear = enqueue(queue, rear, "Mina")
    output_queue(queue, front, rear)

    # Dequeue all of the items
    for item in queue:
        item, front = dequeue(queue, front, rear)
        print(f"\nDequeued {item}")

    # This shouldn't enqueue Eirini since there is no implementation for
    # writing over any dequeued elements
    rear = enqueue(queue, rear, "Eirini")
    

# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()

File: data-structures/queues/c-sharp/queue.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{
    class Queues
    {
        public const int MaxSize = 4;

        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            Console.WriteLine("### Queue ###");

            // Initialise the queue and pointers
            string[] queue = new string[MaxSize];
            int front = 0;
            int rear = -1;

            // Insert test data into the queue
            rear = Enqueue(queue, rear, "Julie");
            OutputQueue(queue, front, rear);
            
            rear = Enqueue(queue, rear, "Rey");
            OutputQueue(queue, front, rear);
            
            rear = Enqueue(queue, rear, "Habib");
            OutputQueue(queue, front, rear);
            
            rear = Enqueue(queue, rear, "Sabrina");
            OutputQueue(queue, front, rear);

            // Trying to enqueue - should say queue is full
            rear = Enqueue(queue, rear, "Mina");
            OutputQueue(queue, front, rear);

            // Dequeue all of the items
            foreach (string item in queue) {
                // Tuple is used to store multiple return values
                Tuple<string, int> dequeuedValues; // item, front

                dequeuedValues = Dequeue(queue, front, rear);
                front = dequeuedValues.Item2;
                Console.WriteLine($"\nDequeued {dequeuedValues.Item1}");
            }

            // This shouldn't enqueue Eirini since there is no implementation for
            // writing over any dequeued elements
            rear = Enqueue(queue, rear, "Eirini");
        }

        // Check if the queue is empty
        public static bool IsEmpty(int front, int rear)
        {
            if (front > rear)
                return true;
            else
                return false;
        }

        // Check if the queue is full
        public static bool IsFull(int rear)
        {
            if (rear + 1 == MaxSize)
                return true;
            else
                return false; 
        }

        // Enqueue an item
        public static int Enqueue(string[] queue, int rear, string data)
        {
            if (IsFull(rear) == true)
                Console.WriteLine($"\nQueue is full - {data} not added");
            else {
                rear = rear + 1;
                queue[rear] = data;
            }
            return rear;
        }

        // Dequeue an item
        public static Tuple<string, int> Dequeue(string[] queue, int front, int rear)
        {
            string dequeuedItem;

            if (IsEmpty(front, rear) == true) {
                Console.WriteLine("\nQueue is empty - nothing to dequeue");
                dequeuedItem = "";
            }
            else {
                dequeuedItem = queue[front];
                front = front + 1;
            }

            return Tuple.Create(dequeuedItem, front);
        }

        // Output the state of the queue
        public static void OutputQueue(string[] queue, int front, int rear)
        {
            // Testing
            Console.Write("\nQueue: ");
            Console.WriteLine("[{0}]", string.Join(", ", queue));
            Console.WriteLine($"Front: {front}");
            Console.WriteLine($"Rear: {rear}");
        }

    }
}